sound/soc/sof/ipc4-pcm.c

```diff
@@ -32,6 +34,174 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
 }
 EXPORT_SYMBOL(sof_ipc4_set_pipeline_state);
 
+static u64 ipc4_cc_read(const struct cyclecounter *cc)
+{
+	struct snd_sof_pcm_stream *spcm = container_of(cc, struct snd_sof_pcm_stream, cc);
+
+	return spcm->llp_slot->reading.wclk_l + ((u64)spcm->llp_slot->reading.wclk_u << 32) -
+			spcm->stream_start_wclk;
+}
+
+static void sof_ipc4_timecounter_init(struct snd_sof_dev *sdev, struct snd_sof_pcm_stream *spcm,
+				      u32 wclk)
+{
+	struct cyclecounter *cc = &spcm->cc;
+	u32 maxsec;
+
+	cc->read = ipc4_cc_read;
+	cc->mask = CLOCKSOURCE_MASK(32);
+
+	/* The counter wraps around after 2^32 -1 / wclk seconds. */
+	maxsec = BIT(32) - 1;
+	do_div(maxsec, wclk);
+	clocks_calc_mult_shift(&cc->mult, &cc->shift, wclk, NSEC_PER_SEC, maxsec);
+
+	/* convert 1 cycle to X ns */
+	spcm->accuracy = (1ULL * cc->mult) >> cc->shift;
+	dev_dbg(sdev->dev, "accuracy %d\n", spcm->accuracy);
+}
+
+/*
+ * calculate stream start time and resolution of wclk in dsp
+ * The llp position in memory window is not the real position for a stream in two cases
+ * (1) dai starts before host in multiple pipeline cases. The llp conunter works before
+ *  stream is started.
+ * (2) multiple streams with mixer. The llp counter is same for all streams.
+ *
+ * the solution is using stream_start_offset for each stream to save the llp counter
+ * when that stream is started. stream_start_offset is based on sample size, so need
+ * to convert it to sample count.
+ * the stream position = current llp value  - stream_start_offset
+ *
+ * to calculate stream start time (wclk0), we use math interpolation
+ *	wclk0				wclk1					wclk2
+ * stream_start_offsets*		llp1					llp2
+ *
+ * wclk0 = wclk2 - (wclk1 -wclk2)/(llp1 -llp2) * (llp2  - stream_start_offsets)
+ */
+static void sof_ipc4_wclk_calibration_work(struct work_struct *work)
+{
+	struct sof_ipc4_pipeline_registers *ppl_reg = NULL;
+	struct sof_ipc4_llp_reading_slot *llp_slot = NULL;
+	struct sof_ipc4_fw_registers *fw_reg;
+	struct sof_ipc4_copier *host_copier = NULL;
+	struct sof_ipc4_copier *dai_copier = NULL;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_pcm_stream *spcm;
+	struct snd_sof_dev *sdev;
+	struct snd_sof_dai *dai;
+	u64 stream_start_position;
+	u32 dai_sample_size;
+	u64 pos_pre, wclk_pre;
+	u64 pos_current;
+	u64 wclk_current;
+	u64 wclk, ratio;
+	u32 node_id;
+	int i;
+
+	spcm = container_of(work, struct snd_sof_pcm_stream, wclk_calibration.work);
+	/* find host & dai to locate info in memory window */
+	for_each_dapm_widgets(spcm->list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		if (WIDGET_IS_AIF(swidget->widget->id)) {
+			host_copier = swidget->private;
```

Comment:
how does this work if you have two host streams going into a mixer? Which of the two host_copier would be selected?

Or does this logic only works for simple host-dai pipeline topologies - which would kill the purpose of the start_offset that's meant precisely to handle mixing usages.

Reply:
@bardliao  please correct me.  I tested on my devices for the following pipelines
When stream0 is working only "host0 -> volume0 ->mixin0 ->mixout0->volume1->dai" is included by stream0
And when stream1 is started after stream0, only "host1-> volume1 -> mixin1->mixout0->volume1->dai" is included by stream1.
so there is no problem
````
stream 0 : host0 -> volume0 ->mixin0 ->mixout0->volume1->dai
                                         |
stream 1 : host1-> volume1 -> mixin1-----
````

Reply:
> @bardliao please correct me. I tested on my devices for the following pipelines When stream0 is working only "host0 -> volume0 ->mixin0 ->mixout0->volume1->dai" is included by stream0 And when stream1 is started after stream0, only "host1-> volume1 -> mixin1->mixout0->volume1->dai" is included by stream1. so there is no problem
> 
> ```
> stream 0 : host0 -> volume0 ->mixin0 ->mixout0->volume1->dai
>                                          |
> stream 1 : host1-> volume1 -> mixin1-----
> ```

Yes, that's what it supposed to be.

Reply:
The problem is that you have a loop on all the PCM devices and you find a single 'host_copier' but you have two in your own example @RanderWang

Reply:
````
for_each_dapm_widgets(spcm->list, i, widget)
````
I only check the current stream, not all